block-layer/content-addressable-archives.md

```diff
@@ -30,51 +30,110 @@
 
 ```
 |--------- Header --------| |---------------------------------- Data -----------------------------------|
 
 [ varint | dag-cbor block ] [ varint | CID | block ] [ varint | CID | block ] [ varint | CID | block ] …
 ```
 
-### Header
+As with Version 2, the initial varint is read then the first block is read from the number of bytes determined by the initial varint. This block is the CAR _Header_, a DAG-CBOR (CBOR with tag `42` for CIDs) encoded object holding the version number and array of roots.
 
-The first bytes of the CAR format hold a varint, this unsigned integer specifies the number of bytes beyond the varint itself that contain the _Header_ block. This Header block is a byte array DAG-CBOR (CBOR with tag 42 for CIDs) encoded object holding the version number and array of roots. As an [IPLD Schema](../schemas/):
+Described as an [IPLD Schema](../schemas/):
 
 ```ipldsch
-type CarHeader struct {
+type CarV1Header struct {
   version Int
   roots [&Any]
 }
 ```
 
+All trailing blocks constitute the Data section of the CAR, where blocks are prefixed by a varint describing the number of bytes to read, followed by the block's CID, followed by the block's binary data.
+
 #### Constraints
 
-* The `version` is always a value of `1`. Future iterations of this specification may make use of `version` to introduce variations of the format.
-* The `roots` array must contain **one or more** CIDs, each of which should be present somewhere in the remainder of the CAR.
+* The `roots` array should contain **one or more** CIDs, each of which should be present somewhere in the remainder of the CAR. A zero-length `roots` array is possible but not recommended due to assumptions present in downstream code.
+
+### Version 2
 
-_(Caveats: see [Number of roots](#number-of-roots) and [Root CID block existence](#root-cid-block-existence) under Unresolved Issues.)_
+```
+|------ Version ------| |----------- Header ------------| |---------------------- Data ----------------------|
+
+[ varint | cbor block ] [ varint | CID | dag-cbor block ] [ varint | CID | block ] [ varint | CID | block ] …
+```
+
+To maintain backward compatibility, Version 2 retains the lead-in block without CID but _only_ includes a CBOR encoded map containing a single key/value pair: `{ version: 2 }`. Therefore, a decoder may read the first 11 bytes of a CAR and match it against the bytes: `0x0aa16776657273696f6e02` (`0x0a` is a varint representing `10`, followed by 10 bytes of the CBOR encoded form of `{ version: 2 }`).
+
+Described as an IPLD Schema:
+
+```ipldsch
+type CarVersion struct {
+  version Int
+}
+```
+
+All following blocks are prefixed by a varint describing the number of bytes to read, followed by the block's CID, followed by the block's binary data.
+
+The first block in this list is the _Header_, a DAG-CBOR encoded object, containing roots, selectors and deterministic descriptor.
+
+Described as an IPLD Schema:
+
+```ipldsch
+type CarRoot struct {
```

Comment:
One thing I want to explore is making the entire CAR header block a completely valid Selector (or perhaps we need to define some notion of a “multi-selector”).

It would be pretty amazing to be able to take a selector block, drop it into a selector engine, and spit out a CAR file with a CID for the header block that matches the selector you passed.

This may not be worth it. For instance, `deterministic` boolean would become problematic as it’s unlikely to be there if you just create a selector block.

Reply:
that would require a selector to ship with a CID embedded in it wouldn't it? I don't think we have that kind of structure in selectors as they are today (afaik!)